webui/react/src/stores/experiments.tsx

```diff
@@ -1,24 +1,176 @@
 import { Loadable, Loaded, NotLoaded } from 'hew/utils/loadable';
 import { Map } from 'immutable';
+import * as t from 'io-ts';
 
+import { valueof } from 'ioTypes';
 import { getExperiments } from 'services/api';
-import { V1Pagination } from 'services/api-ts-sdk';
+import { Jobv1State } from 'services/api-ts-sdk';
 import { GetExperimentsParams } from 'services/types';
-import { ExperimentItem, ExperimentPagination } from 'types';
-import { Observable, observable, WritableObservable } from 'utils/observable';
+import {
+  CheckpointStorageType,
+  ExperimentItem,
+  ExperimentPagination,
+  ExperimentSearcherName,
+  HyperparameterBase,
+  Hyperparameters,
+  HyperparameterType,
+  Json,
+  JsonObject,
+  RunState,
+} from 'types';
+import asValueObject, { ValueObjectOf } from 'utils/asValueObject';
+import { immutableObservable, Observable } from 'utils/observable';
 import { encodeParams } from 'utils/store';
 
 import PollingStore from './polling';
 
-type ExperimentCache = {
-  experimentIds: Readonly<number[]>;
-  pagination: Readonly<V1Pagination>;
-};
+// TODO: use these codecs as source for types when possible
+const jsonCodec: t.Type<Json> = t.recursion('Json', () =>
```

Comment:
Can you put these in `types.ts` and use `t.typeOf` for the json definition.

Reply:
I can move these to types.ts, but io-ts can't infer from recursive types, so the typescript definitions have to stay --  is that okay?

Reply:
That's fine, they match up right?

Reply:
done!